chore(ci): build RPM package

ReactorScram opened the pull request:
Refs #6145 

This bundles aarch64 and x86_64 RPMs in CI and CD.

We'll need a 2nd PR to add everything to the changelog and knowledge base, after the first release with RPMs is cut.

ReactorScram committed:
check in some files

ReactorScram committed:
checkpoint

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into chore/build-rpm

ReactorScram committed:
try building rpm in CI

ReactorScram committed:
fix specfile

ReactorScram committed:
try to fix this path

ReactorScram committed:
rip the biggest libs from Ubuntu

ReactorScram committed:
this one works with only a little manual setup on CentOS 9

ReactorScram committed:
fix arch-specific files

ReactorScram committed:
fixes

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into chore/build-rpm

ReactorScram committed:
fix

ReactorScram committed:
debug

ReactorScram committed:
fix

ReactorScram committed:
fix x86_64

ReactorScram commented on .github/actions/setup-tauri-v2/action.yml:
```diff
@@ -15,7 +15,7 @@ runs:
     - name: Install Tauri build deps
       if: ${{ runner.os == 'Linux' }}
-      run: sudo apt-get --yes install build-essential curl file libwebkit2gtk-4.1-dev libayatana-appindicator3-dev librsvg2-dev libssl-dev libxdo-dev wget
+      run: sudo apt-get --yes install build-essential curl file libwebkit2gtk-4.1-dev libayatana-appindicator3-dev librsvg2-dev libssl-dev libxdo-dev rpm wget
```
Despite ChatGPT's warnings, it seems easy to build an rpm from within Ubuntu.

ReactorScram commented on rust/gui-client/src-tauri/tauri.conf.json:
```diff
@@ -13,18 +13,11 @@
     "targets": [
       "deb",
-      "msi",
-      "rpm"
```
Gave up on Tauri's RPM bundler

ReactorScram committed:
revert some unneeded changes

ReactorScram committed:
add instructions

thomaseizinger commented on rust/gui-client/README.md:
```diff
@@ -75,6 +75,23 @@ If that doesn't work:
+These instructions will move to the knowledge base once the first release supporting CentOS 9 is cut.
+
+1. Download RPM
+2. `sudo dnf install systemd-resolved` (Installing it explicitly prevents it from being auto-removed if Firezone is removed)
```
I think this can be interesting to point out but seems like a decision the user should make? On Debian-based distros, we just auto-install it, no?

ReactorScram replied:
On Ubuntu we assume it to already be installed. I got a shock on CentOS when I un-installed Firezone and `/etc/resolv.conf` was still pointing to systemd-resolved, but it had apparently un-installed systemd-resolved too, so my DNS was broken. (I think apt would do the same in this situation)

thomaseizinger replied:
Yeah probably. I'd expect the systemd-resolved uninstaller to fix the symlink :(

thomaseizinger commented on rust/gui-client/README.md:
```diff
@@ -75,6 +75,23 @@ If that doesn't work:
+
+1. Download RPM
+2. `sudo dnf install systemd-resolved` (Installing it explicitly prevents it from being auto-removed if Firezone is removed)
+3. `sudo dnf install epel-release` (Needed to get GNOME extensions)
```
Why can't we express this as a dependency?

thomaseizinger replied:
I guess same reason as below, might not be using Gnome.

thomaseizinger commented on rust/gui-client/README.md:
```diff
@@ -75,6 +75,23 @@ If that doesn't work:
+5. `sudo usermod -aG firezone-client $USER`
+6. `sudo systemctl enable firezone-client-ipc.service` (See https://www.freedesktop.org/software/systemd/man/latest/systemd.preset.html, "It is not recommended to ship preset files within the respective software packages implementing the units")
+7. `sudo systemctl enable systemd-resolved.service` (In addition to the preset thing, the family of Fedora distros seem to have a policy that installing a service should never enable or auto-start it. This might be contrary to Debian-family.)
+8. `sudo dnf install gnome-shell-extension-appindicator` (Install the system tray)
```
Why can't this be a dependency?

ReactorScram replied:
They might not be running GNOME, right?

thomaseizinger replied:
Oh, we are accounting for that. Yeah makes sense.

thomaseizinger replied:
Can we point this out then? Like "install system tray of your desktop, e.g. ..."

thomaseizinger commented on rust/gui-client/README.md:
```diff
@@ -75,6 +75,23 @@ If that doesn't work:
+4. `sudo dnf install ./firezone-client-gui-*.rpm`
+5. `sudo usermod -aG firezone-client $USER`
+6. `sudo systemctl enable firezone-client-ipc.service` (See https://www.freedesktop.org/software/systemd/man/latest/systemd.preset.html, "It is not recommended to ship preset files within the respective software packages implementing the units")
+7. `sudo systemctl enable systemd-resolved.service` (In addition to the preset thing, the family of Fedora distros seem to have a policy that installing a service should never enable or auto-start it. This might be contrary to Debian-family.)
```
Does our IPC service set a dependency to this? If yes, we don't necessarily need to enable it separately, do we?

If somebody wanted to use systemd-resolved separately, then this would be useful but I am not sure we need to do this as part of our setup?

ReactorScram replied:
Added that as an `After` and `Want` so systemd will auto-start `systemd-resolved` for us even though it's not enabled, so that's really nice, saves a step.

thomaseizinger commented on rust/gui-client/README.md:
```diff
@@ -75,6 +75,23 @@ If that doesn't work:
+2. `sudo dnf install systemd-resolved` (Installing it explicitly prevents it from being auto-removed if Firezone is removed)
+3. `sudo dnf install epel-release` (Needed to get GNOME extensions)
+4. `sudo dnf install ./firezone-client-gui-*.rpm`
+5. `sudo usermod -aG firezone-client $USER`
```
The sys-users stuff from systemd doesn't work on rpm based distros?

ReactorScram replied:
It works, it creates the group, but this step is also needed on Ubuntu. It's the same as adding yourself to the `docker` group, I don't think installing a package is supposed to add users to groups

thomaseizinger replied:
Ah yes, got it mixed up!

thomaseizinger commented on rust/gui-client/README.md:
```diff
@@ -75,6 +75,23 @@ If that doesn't work:
+7. `sudo systemctl enable systemd-resolved.service` (In addition to the preset thing, the family of Fedora distros seem to have a policy that installing a service should never enable or auto-start it. This might be contrary to Debian-family.)
+8. `sudo dnf install gnome-shell-extension-appindicator` (Install the system tray)
+9. Reboot to add user to group fully, start new services, and make the appindicator extension available.
+10. `sudo ln --force --symbolic /run/systemd/resolve/stub-resolv.conf /etc/resolv.conf`
```
Is that because we try to read resolv.conf and it isn't there otherwise on these distros?

ReactorScram replied:
~~If~~ When I do another test run I'll double-check, but I think systemd-resolved just doesn't try to control this file. As admin we have to switch it over

ReactorScram replied:
Looks like systemd-resolved does not try to control it even when enabled and started. I noted in the docs that it's probably because it's under `/etc` so it would be against policy for a package to automatically change it. 🤷‍♀️

thomaseizinger commented on rust/gui-client/README.md:
```diff
@@ -75,6 +75,23 @@ If that doesn't work:
+8. `sudo dnf install gnome-shell-extension-appindicator` (Install the system tray)
+9. Reboot to add user to group fully, start new services, and make the appindicator extension available.
+10. `sudo ln --force --symbolic /run/systemd/resolve/stub-resolv.conf /etc/resolv.conf`
+11. `gnome-extensions enable appindicatorsupport@rgcjonas.gmail.com` (Yes, really. Also, this will tab-complete.)
```
This only works for a reboot?

ReactorScram replied:
Yes, really. It's bizarre but I got errors saying "appindicatorsupport@... doesn't exist" and then rebooted and it worked fine.

ReactorScram replied:
Actually logging out and back in to restart the GNOME session is probably enough. But IME for `usermod` to finish its work, a reboot is cleaner. Logging out and back in doesn't seem to add a user to a group fully.

(Will also check this on the final test run. Ubuntu needed a reboot but it would be nice to not ask users to reboot)

ReactorScram replied:
Was able to weaken this to a log-out-and-log-in cycle

thomaseizinger reviewed (approved):
Looks painful to get to the bottom of ... 

Thanks for digging through out!
I think a .desktop file would be useful.

ReactorScram committed:
add desktop file and icons to RPM

ReactorScram committed:
redo instructions, add systemd-resolved as a want

ReactorScram committed:
correct version

ReactorScram committed:
restart the tunnel daemon on upgrades

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into chore/build-rpm

ReactorScram committed:
remove SHA256 pin of CentOS

I only started on this a couple days ago and it's already out of date, so I'm guessing it's constantly updating.

ReactorScram committed:
remember to use stream9

ReactorScram committed:
can't be interactive in CI ofc

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into chore/build-rpm

ReactorScram committed:
un-mess-up perms

ReactorScram committed:
try again
